### PR TITLE
Form - New tests

### PR DESCRIPTION
**Pull request by novsstation**
1) rename "form.scenarios.tests" to  "form.layout.tests"
2) new test for updateFormData logic and editor values after rendering

### CODE REVIEW

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js [hunk 1]
```diff
@@ -329,3 +329,165 @@ QUnit.test('Reset editor\'s value when set formData: {dataField1: a}', function(
         assert.deepEqual(editor.option('value'), editor._getDefaultOptions().value, `a default value of the ${dataField} editor`);
     });
 });
+
+QUnit.module('UpdateFormData for Textbox editor field', () => {
+    ['2', null, undefined, 'no member'].forEach(oldValue => {
+        ['3', null, undefined, 'no member'].forEach(newValue => {
+            QUnit.test(`FormData = { a: '1', b: ${oldValue} } -> UpdateFormData({ b: ${newValue} })`, function(assert) {
+                const oldFormData = { a: '1' };
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:337]: Do we really need to test it for the full set of configs for each widget?
imho it's enough to test one member only: old -> new.

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js [hunk 2]
```diff
@@ -329,3 +329,165 @@ QUnit.test('Reset editor\'s value when set formData: {dataField1: a}', function(
...
+                const form = $('#form').dxForm({
+                    formData: oldFormData,
+                    items: [
+                        { dataField: 'a', editorType: 'dxTextBox' },
+                        { dataField: 'b', editorType: 'dxTextBox' }
+                    ],
+                }).dxForm('instance');
+                const expectedOldValue = (oldValue === 'no member' || oldValue === undefined) ? '' : oldValue;
+                assert.strictEqual(form.getEditor('a').option('value'), '1');
+                assert.strictEqual(form.getEditor('b').option('value'), expectedOldValue);
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:351]: I think we need to check the formData.member value as well:
assert.strictEqual(form.option('formData').member, expectedOldValue);  - it should be synchronized with the editor value

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js [hunk 3]
```diff
@@ -329,3 +329,165 @@ QUnit.test('Reset editor\'s value when set formData: {dataField1: a}', function(
...
+
+                const newFormData = {};
+                if(newValue !== 'no member') {
+                    newFormData['b'] = newValue;
+                }
+
+                form.updateData(newFormData);
+                const expectedNewValue = newValue === 'no member' ? expectedOldValue : newValue;
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:359]: Wow...
The '2' value is expected in the editor if the {b: '2'} object was changed to the {} object... Am I right?

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js [hunk 4]
```diff
@@ -329,3 +329,165 @@ QUnit.test('Reset editor\'s value when set formData: {dataField1: a}', function(
...
+                assert.strictEqual(form.getEditor('a').option('value'), '1');
+                assert.strictEqual(form.getEditor('b').option('value'), expectedNewValue);
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:361]: I think we need to check the formData.member value as well:
assert.strictEqual(form.option('formData').member, expectedNewValue); - it should be synchronized with the editor value

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js [hunk 5]
```diff
@@ -329,3 +329,165 @@ QUnit.test('Reset editor\'s value when set formData: {dataField1: a}', function(
...
+            });
+        });
+
+        QUnit.test(`FormData = { a: '1', b: ${oldValue} } -> option('formData', {})`, function(assert) {
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:365]: It seems that this config duplicates the '['no member'].forEach(newValue => {' config and we can delete it.

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js [hunk 6]
```diff
@@ -329,3 +329,165 @@ QUnit.test('Reset editor\'s value when set formData: {dataField1: a}', function(
...
+            assert.strictEqual(form.getEditor('b').option('value'), '');
+        });
+    });
+});
+
+QUnit.module('UpdateFormData for Checkbox editor field', () => {
+    function checkCheckboxEditor(editor, value) {
+        QUnit.assert.strictEqual(editor.option('value'), value, `editor has ${value} value`);
+
+        const $checkBox = $(editor.element());
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:389]: "$checkBox.hasClass" seems to be internal implementation details of dxCheckBox and we should not check it here.

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js [hunk 7]
```diff
@@ -329,3 +329,165 @@ QUnit.test('Reset editor\'s value when set formData: {dataField1: a}', function(
...
+            form.option('formData', {});
+            assert.strictEqual(form.getEditor('a').option('value'), '');
+            assert.strictEqual(form.getEditor('b').option('value'), '');
+        });
+    });
+});
+
+QUnit.module('UpdateFormData for Checkbox editor field', () => {
+    function checkCheckboxEditor(editor, value) {
+        QUnit.assert.strictEqual(editor.option('value'), value, `editor has ${value} value`);
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:387]: I think we need to check the formData.member value as well:
assert.strictEqual(form.option('formData').member, value); - it should be synchronized with the editor value

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js [hunk 8]
```diff
@@ -329,3 +329,165 @@ QUnit.test('Reset editor\'s value when set formData: {dataField1: a}', function(
...
+                }
+
+                form.updateData(newFormData);
+                const expectedNewValue = newBoolValue === 'no member' ? expectedOldValue : newBoolValue;
+                checkCheckboxEditor(form.getEditor('innerObject.a'), true);
+                checkCheckboxEditor(form.getEditor('innerObject.b'), expectedNewValue);
+            });
+        });
+
+        QUnit.test(`FormData = { a: true, b:  ${oldBoolValue}} -> option('formData', {})`, function() {
```
Comment (IgnatovDan) [inline testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:455]: This config seems to be covered by

    [true, false, undefined, null, 'no member'].forEach(oldBoolValue => {
        ['no member'].forEach(newBoolValue => {

Remove?

**Review by IgnatovDan — CHANGES_REQUESTED** [review]

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js (hunk 2 above)
Comment (novsstation) [reply testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:351]: Fixed in https://github.com/DevExpress/DevExtreme/pull/13450/files

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js (hunk 4 above)
Comment (novsstation) [reply testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:361]: Fixed in https://github.com/DevExpress/DevExtreme/pull/13450/files

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js (hunk 6 above)
Comment (novsstation) [reply testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:389]: It is jQuery method. 
I have check the existence of the dx-check-box-checked and dx-checkbox- intermediate classes to verify visual representation of the checkbox

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js (hunk 5 above)
Comment (novsstation) [reply testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:365]: Fixed in https://github.com/DevExpress/DevExtreme/pull/13450/files

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js (hunk 3 above)
Comment (novsstation) [reply testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:359]: Yep.
If it does not exists as field of the parameter of the updateData method? the the value does't change.

File: testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js (hunk 1 above)
Comment (novsstation) [reply testing/tests/DevExpress.ui.widgets.form/form.API.option_formData.tests.js:337]: Fixed in https://github.com/DevExpress/DevExtreme/pull/13450/files

**Pull request closed without merging**